### ISSUE TITLE
Missing memory probing

### ISSUE THREAD

**Issue by steveeJ** [intermezzOS/kernel#114]
In order to manage the memory, the OS should know how much it has. Extra information about reserved addresses and memory holes (maybe these are the same, I don't know yet), should also be taken into account.

**Comment by steveklabnik** [comment]
Yup, this is absolutely something that would have to happen for https://github.com/intermezzOS/kernel/issues/96 and https://github.com/intermezzOS/kernel/issues/81

**Comment by dato** [comment]
An expedient option would be to get the information from the bootloader. The multiboot specification includes information on [how much memory the system has](http://wiki.osdev.org/Detecting_Memory_(x86)#Memory_Map_Via_GRUB).

This would make the dependency on grub/multiboot a hard one, though; but might be a good start.

**Comment by steveklabnik** [comment]
@dato big 👍 on that, multiboot should have all the info. I am fine with this.

**Comment by steveeJ** [comment]
@dato do you by chance know how this is handled in (U)EFI systems?

**Comment by dato** [comment]
I’m still using BIOS for my x86_64 OS work, but it’s a similar
story as Multiboot: it’s handed to you directly (in this case by
EFI firmware), no need to do a special dance with BIOS services
and interrupts to get it.

You want, in particular, the ‘GetMemoryMap()’ service. There is
example code in this [excellent StackOverflow answer]. For those
reading along, here’s an [overall description].

-d


[excellent StackOverflow answer]
https://stackoverflow.com/a/17592211/848301

[overall description]
https://hypervsir.blogspot.com.ar/2014/09/approach-to-retrieving-bios-memory-map.html

**Comment by dato** [comment]
I’m still using BIOS for my x86_64 OS work, but it’s a similar
story as Multiboot: it’s handed to you directly (in this case by EFI firmware), no need to do a special dance with BIOS services and interrupts to get it.

You want, in particular, the `GetMemoryMap()` service. There is example code in this [excellent StackOverflow answer](https://stackoverflow.com/a/17592211/848301). For those reading along, here’s an [overall description](https://stackoverflow.com/a/17592211/848301).

-d
